src/bin/pgcopydb/ld_apply.c

```diff
@@ -495,13 +495,30 @@ stream_apply_sql(StreamApplyContext *context,
 		case STREAM_ACTION_BEGIN:
 		{
 			if (metadata->lsn == InvalidXLogRecPtr ||
-				metadata->txnCommitLSN == InvalidXLogRecPtr ||
 				IS_EMPTY_STRING_BUFFER(metadata->timestamp))
 			{
 				log_fatal("Failed to parse BEGIN message: %s", sql);
 				return false;
 			}
 
+			/* if txnCommitLSN is be invalid, then fetch it from txn metadata file */
+			if (metadata->txnCommitLSN == InvalidXLogRecPtr)
+			{
+				log_trace("stream_apply_sql: BEGIN message without commit LSN, "
+						  "fetching commit LSN from transaction metadata file");
```

Comment:
This looks like it should be LOG_DEBUG, as it's not an implementation detail. I try to keep TRACE level for checking internals (algorithms, resource usage, low-level programmer things). Again, there is no hard rule about it, unfortunately. Here the user can follow what's happening. I would go as far as:

  1. write a separate function to compute the filename where to read/parse the XID from
  2. use a log_debug message that includes the name of the file, so that if something goes wrong we can use the logs to open the file and see its contents and if it exists etc
  3. change `parseTxnMetadataFile` to take a `const char *filename, LogicalMessageMetadata *metadata` argument list